types/sanctuary/index.d.ts

```diff
@@ -107,19 +103,25 @@ declare namespace Sanctuary {
     gte<A>(x: Ord<A>): (y: Ord<A>) => boolean;
     min<A>(x: Ord<A>): (y: Ord<A>) => A;
     max<A>(x: Ord<A>): (y: Ord<A>) => A;
+    clamp(tk: TK): (tk: TK) => (tk: TK) => TK;
```

Comment:
I think it goes the same way as examples above:
```
clamp<A>(min: A): (max: A) => (value: A) => A;
```
But I dislike the way `Ord` is typed. I think it should be done with proper interface and union type (just like `Show`), so it allows use of custom types:
```
interface Ord<A> {
  compareTo(a: A, b: A): boolean; // or whatever underlying implementation uses
}

clamp<T extends Ord<T> | number | string | Date | etc>(min: T): (max: T) => (value: T) => T;
```

Reply:
I would love to see this work. Attempting to write TypeScript type definitions for Sanctuary has been a source of frustration for me over the past few years; I hope we have more success this time. The crucial thing is to prohibit `S.clamp ('abc') (123)`, which is invalid due to a type-variable constraint violation even though the Ord constraint is satisfied in both cases.

Reply:
Nothing simpler: [typescript playground example](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgPJQCYB4CCA+ZAbwChlkEB7AWwAc4oIAVCgCgrAAtoAuZHASl4AjChQA2EOCADcyAPRzkFKMgDuHOJABu0ZAFcQGaGICeoAObJgtCVQjhNwCiH0BnCK+IBfYsRgGEMCcXBDE4WixGZAgAD0hDVzRMSIIAH2QQPSohXXSAEU0UdNcwKAs8FjheRgAaZCFquq04MT0IasFkKJIyBjA9KBdm1ohZBWQYZWRXOABrFAoYaesaMWAEYDATb19xhD0S6mQtmgg69XWOZA0aU5BEsAorGwg7cGRzYB0XUEhYRAgxFCcFciQAYqIiKRyNQ6AxmGxODxkBCKPwoWRehB+oMJi13NJoT4fECwrQWAAmOoAcjgQgQ1LqFP4pPCNBYIAgqhRohY-DqAEYKQBmOoAFhZQA)

Reply:
Great! The next steps are to see whether it works with curried functions, and with a property named `fantasy-land/lte` rather than `compareTo`.

Reply:
No problem at all: [another playground](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMoTAe2AEwDwCCAfMgN4BQyyA5POHAM4CeAtADZwg4D0EAjgFc4bBtQBcyABSYwAC2gSCASmQBeEgCNMmNhE4BucgF9y5UJFiIUAeSj5iyCAA9IXBmgzZ7JClVqcwRlYOLm42SHEpGXkoRRV1ZC0dPRBDE3IETBAGMGQEDgBbAAc1ZDwAFUcXCDdkW3xykgAfZBABAo1oZBaAEThIbuQcqFAAcyJJOAly+JJJDWnZqQA3YQEIRbUiSmRVtnXkbm5kGEwoIbgAaxRMGCHgYrZgBGAwJlMjvIEczALkN6KEAANMgAO6yZ6yZCyOBFQHZf6YZAPIq6Ao1XKjYDLGrI8DQeBIDIcBjuABi2jIO389GY7E4PHCEGo0jkCmQFMwKl8VGQUAwAigIH+UHWhio6T8dECdJCPH4QhELOi7M53J2VH5YEFwrAoog4uQJnS+TgxUkACYlJJqHANAhqNarcSzUVJCAIKCOdpJEprQBGC0AZmtABYlOQgA). I am personally convinced that _all_ interfaces which are currently empty should be written this way. To be honest, theirs current (empty) implementation is misconceived from the ground up to me, as "type constraints" you mentioned doesn't work at all, and it disallows custom types from being used. This misconception might've been the root of whole problem in the first place.

Reply:
This is very promising. Let's update the interfaces as you suggest. :)